### PR TITLE
DX: move all dev dependencies to dev-tools

### PR DESCRIPTION
**Pull request by kubawerlos**
This PR extracts `dev-require` and `dev-autoload` from `composer.json`, but also:

- cleans up AppVeyor, CircleCI and TravisCI configurations
- update TravisCI fast test to run PHP 7.2 instead of PHP 7.0
- updates `InstallViaComposerTest`, so it makes the copy for testing, not symlink

Should those 3 be extracted to separate PRs?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -60,46 +60,73 @@ jobs:
...
                 - PHP_CS_FIXER_FUTURE_MODE=1 php php-cs-fixer --diff --dry-run -v fix
 
         -
             <<: *STANDARD_TEST_JOB
             stage: Test
-            php: 5.6
+            php: '5.6'
             env: COMPOSER_FLAGS="--prefer-stable --prefer-lowest"
+            before_install:
+                # turn off XDebug
```
Comment (localheinz) [inline .travis.yml]: See https://twitter.com/derickr/status/1184465999971700736.

😉

Comment (kubawerlos) [reply .travis.yml]: @localheinz fixed, now you can safely review the rest of PR ;)

File: .appveyor.yml [hunk 2]
```diff
@@ -35,11 +35,12 @@ install:
 
 before_test:
     - cd C:\projects\php-cs-fixer
-    - php C:\tools\composer.phar update --optimize-autoloader --no-interaction --no-progress --prefer-stable --no-ansi
-    - php C:\tools\composer.phar info -D | sort
+    - php C:\tools\composer.phar remove --dev --no-update --working-dir=./dev-tools humbug/box localheinz/composer-normalize phpstan/phpstan-phpunit
```
Comment (keradus) [inline .appveyor.yml]: why `remove` ?

File: .travis.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ jobs:
     include:
         -
             stage: Static Code Analysis
-            php: 7.3
+            php: '7.3'
```
Comment (keradus) [inline .travis.yml]: please revert

File: php-cs-fixer [hunk 4]
```diff
@@ -56,7 +56,7 @@ if (class_exists('Phar')) {
 
 if ($require) {
     // OK, it's not, let give Composer autoloader a try!
-    $possibleFiles = [__DIR__.'/../../autoload.php', __DIR__.'/../autoload.php', __DIR__.'/vendor/autoload.php'];
+    $possibleFiles = [__DIR__.'/../../autoload.php', __DIR__.'/../autoload.php', __DIR__.'/vendor/autoload.php', __DIR__.'/dev-tools/vendor/autoload.php'];
```
Comment (keradus) [inline php-cs-fixer]: it is a prod file. it must not be aware about `dev-tools`

File: .appveyor.yml (hunk 2 above)
Comment (kubawerlos) [reply .appveyor.yml]: One job runs on PHP 5.6 and these packages need 7+.

File: .travis.yml (hunk 3 above)
Comment (kubawerlos) [reply .travis.yml]: why? PHP version is not a float value, but a string. PHP 7.0 is not `7.0`, but `7` on the [list](https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/builds/613530218).

File: composer.json [hunk 5]
```diff
@@ -31,20 +31,6 @@
...
-    "require-dev": {
-        "johnkary/phpunit-speedtrap": "^1.1 || ^2.0 || ^3.0",
-        "justinrainbow/json-schema": "^5.0",
-        "keradus/cli-executor": "^1.2",
-        "mikey179/vfsstream": "^1.6",
-        "php-coveralls/php-coveralls": "^2.1",
-        "php-cs-fixer/accessible-object": "^1.0",
-        "php-cs-fixer/phpunit-constraint-isidenticalstring": "^1.1",
-        "php-cs-fixer/phpunit-constraint-xmlmatchesxsd": "^1.1",
-        "phpunit/phpunit": "^5.7.27 || ^6.5.14 || ^7.1",
```
[the remark this answers is not in the document]
Comment (kubawerlos) [reply composer.json]: why "must"? what is the problem if they are not?

File: php-cs-fixer (hunk 4 above)
Comment (kubawerlos) [reply php-cs-fixer]: I agree that is the perfect solution.

Yet I can name [one file](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.0/src/Console/Command/HelpCommand.php#L287) that is aware of `dev-tools`.

Also, what harm can it make? Only if previous 3 paths are not matching that last one is even considered - and I don't believe would match for anyone.

In fact, only one test class was failing without this change - I can give it another try if you insist.

File: .travis.yml (hunk 3 above)
Comment (keradus) [reply .travis.yml]: i don't see anyhow this argument.
even if string, then `php: 7.4snapshot` and not `php: '7.4snapshot'`.
same, the `stage` is a string, but we don't use quotes. they are not needed

File: composer.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (keradus) [reply composer.json]: that is a purpose of `require-dev`.
we are extending classes from phpunit package in our codebase - thus it must be there in composer.json file. if we use the class in src - then package should be in `dependencies`. if not, then in `devdependencies`.

removing dependencies directly used in code (dev or src) from composer file i consider a very bad practice. it is not the external tool that we are using out of code, like phpmd.

File: .travis.yml [hunk 6]
```diff
@@ -60,46 +60,73 @@ jobs:
...
             <<: *STANDARD_TEST_JOB
             stage: Test
-            php: 7.2
+            php: '7.1'
+            before_install:
+                # turn off Xdebug
+                - phpenv config-rm xdebug.ini || return 0
+
+                # Composer: boost installation
+                - composer global show hirak/prestissimo --quiet || travis_retry composer global require $DEFAULT_COMPOSER_FLAGS hirak/prestissimo
```
Comment (keradus) [inline .travis.yml]: DRY is violated

File: composer.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (keradus) [reply composer.json]: maybe you should just suggest to Composer to drop the `require-dev` section at all?

File: php-cs-fixer (hunk 4 above)
Comment (keradus) [reply php-cs-fixer]: Point existing code smell is not the best excuse to introduce another...

i consider referencing to dev file available in dev env only to be a code smell

File: tests/Smoke/InstallViaComposerTest.php [hunk 7]
```diff
@@ -75,6 +75,9 @@ public function testInstallationViaPathIsPossible()
                 [
                     'type' => 'path',
                     'url' => __DIR__.'/../..',
+                    'options' => [
+                        'symlink' => false,
```
Comment (keradus) [inline tests/Smoke/InstallViaComposerTest.php]: any reasoning ?

File: .appveyor.yml (hunk 2 above)
Comment (keradus) [reply .appveyor.yml]: yet, on stable branch, there is no issue like that in AppVeyor
https://ci.appveyor.com/project/keradus/php-cs-fixer/builds/29070411

Comment (kubawerlos) [reply .appveyor.yml]: Because those packages are already in dev-tools.

File: .travis.yml (hunk 3 above)
Comment (kubawerlos) [reply .travis.yml]: `stage` is a string because it is a string.

I was always thinking we want be explicit over implicit - I guess not in this case.

File: composer.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (kubawerlos) [reply composer.json]: Purpose of `require-dev` is to lists packages required for developing this package - so following that reasoning all dependencies from `dev-tools` should be moved to main `composer.json`.

> if we use the class in src - then package should be in `dependencies`.

Is that rule about the src directory? ;) I believe (no-dev) dependencies are the ones needed in production - which in case of library is release - such as PHPUnit - because [some files](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.16.0/composer.json#L57) relying on it are in the release. Yet, no one ever consider moving it from `require-dev` to `require`.

> removing dependencies directly used in code (dev or src) from composer file i consider a very bad practice

I don't understand why? I mean the part for `dev`. For `src` it's obvious, but for `dev`? Even if PHP CS Fixer would be install via composer locally in project it's dev dependencies would be ignored.

TL;DR the idea was to move out dev dependencies from main `composer.json` to free normal dependencies from any potential conflict, nothing more.

File: php-cs-fixer (hunk 4 above)
Comment (kubawerlos) [reply php-cs-fixer]: > i consider referencing in prod file to dev file available in dev env only to be a code smell

me too, but for me in this case it's a valid exception to do that

File: .travis.yml (hunk 6 above)
Comment (kubawerlos) [reply .travis.yml]: I didn't know there is a rule to not repeat the code more then twice ;)

Do you know how can we fix this and currently existing duplication as well?

File: tests/Smoke/InstallViaComposerTest.php (hunk 7 above)
Comment (kubawerlos) [reply tests/Smoke/InstallViaComposerTest.php]: This test would fail - without it packages was installed via symlink and then calling `php-cs-fixer` would require `autoload.php` from project's root `vendor`, not the one created for the test.

File: php-cs-fixer (hunk 4 above)
Comment (kubawerlos) [reply php-cs-fixer]: on the second thought - and after cleaning the other smell - this needs fixing

File: .travis.yml (hunk 3 above)
Comment (keradus) [reply .travis.yml]: i see no reasoning why `stage` should not have quotes and `php` should
we have, on master line:
```
stage: Static Code Analysis
php: 7.4snapshot
```

both are strings. none is using quotes

File: composer.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (keradus) [reply composer.json]: we don't care about conflicting dependencies of standalone tools, like phpmd. so we extracted it. cool.
we do care about conflicting dependencies of dev libraries we are using in our own codebase, eg phpunit, so let's keep them under main composer.json.

File: .travis.yml (hunk 6 above)
Comment (keradus) [reply .travis.yml]: probably each duplication to discuss separately. i'm talking about multiple places in travis file when you added same step over and over ago. not needed, as proven on master branch already.

File: tests/Smoke/InstallViaComposerTest.php (hunk 7 above)
Comment (keradus) [reply tests/Smoke/InstallViaComposerTest.php]: why it's not failing on master branch when symlink is not there?

Comment (kubawerlos) [reply tests/Smoke/InstallViaComposerTest.php]: it uses original code, so all is good - with these changes there is no `vendor/autoload.php` as all was moved to `dev-tools`
